sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md

```diff
@@ -0,0 +1,776 @@
+expressiveness for explicit SIMD data-parallel programming while enabling
+efficient mapping to SIMD vector operations on Intel GPU architectures.
+
+### SIMD vector class
+
+The `sycl::intel::gpu::simd` class is a vector templated on some element type. The element type must be vectorizable type. The set of vectorizable types is the set of fundamental SYCL arithmetic types (C++ arithmetic types or `half` type) excluding `bool`. The length of the vector is the second template parameter.
+
+ESIMD compiler back-end does the best it can to map each `simd` class object to a consecutive block
+of registers in the general register file (GRF).
+
+```cpp
+namespace sycl {
+namespace intel {
+namespace gpu {
+
+// vector_type, using clang vector type extension.
+template <typename Ty, int N> struct vector_type {
+  static_assert(!std::is_const<Ty>::value, "const element type not supported");
+  static_assert(is_vectorizable_v<Ty>::value, "element type not supported");
+  static_assert(N > 0, "zero-element vector not supported");
+
+  static constexpr int length = N;
+  using type = Ty attribute((ext_vector_type(N)));
+};
+
+template <int N>
+using mask_type_t = typename vector_type<uint16_t, N>::type;
+
+template <typename Ty, int N> class simd {
+public:
+  using value_type = simd<Ty, N, 0>;
+  using element_type = Ty;
+  using vector_type = vector_type_t<Ty, N>;
+  static constexpr int length = N;
+
+  // Constructors.
+  constexpr simd();
+  constexpr simd(const simd &other);
+  constexpr simd(simd &&other);
+  constexpr simd(const vector_type &Val);
+  constexpr simd(std::initializer_list<Ty> Ilist) noexcept;
+  constexpr simd(Ty Val, Ty Step = Ty()) noexcept;
+
+  // Assignment operators.
+  constexpr simd &operator=(const simd &) &;
+  constexpr simd &operator=(simd &&) &;
+
+  // Subscript operators.
+  Ty operator[](int i) const;
+
+  // Unary operators.
+  simd &operator++();
+  simd operator++(int);
+  simd &operator--();
+  simd operator--(int);
+
+  // Binary operators.
+  auto operator +(const simd &RHS) const;
+  auto operator -(const simd &RHS) const;
+  auto operator *(const simd &RHS) const;
+  auto operator /(const simd &RHS) const;
+  simd operator &(const simd &RHS) const;
+  simd operator |(const simd &RHS) const;
+  simd operator ^(const simd &RHS) const;
+
+  // Compound assignment.
+  simd &operator +=(const simd &RHS);
+  simd &operator -=(const simd &RHS);
+  simd &operator *=(const simd &RHS);
+  simd &operator /=(const simd &RHS);
+  simd &operator &=(const simd &RHS);
+  simd &operator |=(const simd &RHS);
+  simd &operator ^=(const simd &RHS);
+
+  // Compare operators.
+  simd<uint16_t, N> operator >(const simd &RHS) const;
+  simd<uint16_t, N> operator >=(const simd &RHS) const;
+  simd<uint16_t, N> operator <(const simd &RHS) const;
+  simd<uint16_t, N> operator <=(const simd &RHS) const;
+  simd<uint16_t, N> operator ==(const simd &RHS) const;
+  simd<uint16_t, N> operator !=(const simd &RHS) const;
```

Comment:
Returning `simd<uint16_t, N>` looks rather target-specific to me. Are you certain this needs no further abstraction to make it more portable?

Reply:
At least it should be `mask_type_t`, which might become more abstracted in the future revisions.

Reply:
added TODO